libs/server/Lua/LuaRunner.Functions.cs

```diff
@@ -2773,6 +2774,813 @@ static bool TryDecodeLargeMap(LuaRunner self, ref ReadOnlySpan<byte> data, out i
+
+                result = result * 10 + digit;
+                optIx++;
+            }
+
+            return result;
+        }
+
+        static int GetToAlign(int len, int align, char opt, int size)
+        {
+            if (size == 0 || opt == 'c')
+                return 0;
+
+            if (size > align)
+                size = align;
+
+            return (size - (len & (size - 1))) & (size - 1);
+        }
+
+        static bool TryParseControlOptions(LuaRunner self, char opt, string format, ref Header h, ref int optIx, out int constStrRegisteryIndex)
+        {
+            const int MAXALIGN = 8;
+            switch (opt)
+            {
+                case ' ':
+                    constStrRegisteryIndex = -1;
+                    return true;
+
+                case '>':
+                    h.Endian = (int)Endian.BIG;
+
+                    constStrRegisteryIndex = -1;
+                    return true;
+
+                case '<':
+                    h.Endian = (int)Endian.LITTLE;
+
+                    constStrRegisteryIndex = -1;
+                    return true;
+
+                case '!':
+                    int a = GetNum(self, format, ref optIx, MAXALIGN);
+                    if (!BitOperations.IsPow2(a))
+                    {
+                        // Alignment {a} is not a power of 2
+                        constStrRegisteryIndex = self.LuaWrappedError(1, self.constStrs.ErrBadArg);
+                        return false;
+                    }
+                    h.Align = a;
+
+                    constStrRegisteryIndex = -1;
+                    return true;
+
+                default:
+                    constStrRegisteryIndex = self.LuaWrappedError(1, self.constStrs.ErrBadArg);
+                    return false;
+            }
+        }
+
+        /// <summary>
+        /// Entry point for struct.pack from a Lua script.
+        /// </summary>
+        internal int StructPack(nint luaStatePtr)
+        {
+            state.CallFromLuaEntered(luaStatePtr);
+
+            var numLuaArgs = state.StackTop;
+            if (numLuaArgs == 0 || !TryGetFormat(this, 1, out var format))
+            {
+                return LuaWrappedError(1, constStrs.BadArgPack);
+            }
+
+            scratchBufferBuilder.Reset();
+
+            // Parse format
+            int totalSize = 0;
+            int optIx = 0;
+            Header h = DefaultOptions();
+
+            while (optIx < format.Length)
+            {
+                char opt = format[optIx++];
+                int size = OptSize(this, opt, format, ref optIx);
+                int toAlign = GetToAlign(totalSize, h.Align, opt, size);
+                totalSize += toAlign;
+
+                while (toAlign-- > 0)
+                {
+                    AddAlignmentPadding(this);
+                }
+
+                switch (opt)
+                {
+                    case 'b': case 'B':
```

Comment:
nit: consistent `case` layout - let's put each `case foo:` on it's own line, like is done with `'c'` and `'s'`.